documentation/doxygen.py

```diff
@@ -1992,39 +1992,69 @@ def parse_func(state: State, element: ET.Element):
     func.brief = parse_desc(state, element.find('briefdescription'))
     func.description, templates, params, func.return_value, func.return_values, func.exceptions, search_keywords, func.deprecated, func.since = parse_func_desc(state, element)
 
-    # Friend functions have friend as type. That's just awful. COME ON.
-    if func.type.startswith('friend '):
-        func.type = func.type[7:]
-
     def is_identifier(a): return a == '_' or a.isalnum()
 
     # Extract function signature to prefix, suffix and various flags. Important
     # things affecting caller such as static or const (and rvalue overloads)
     # are put into signature prefix/suffix, other things to various is_*
     # properties.
-    if func.type == 'constexpr': # Constructors
-        func.type = ''
-        func.is_constexpr = True
-    elif func.type.startswith('constexpr'):
-        func.type = func.type[10:]
-        func.is_constexpr = True
-    # For some effing reason, when a constexpr function has decltype(auto)
-    # return type, Doxygen swaps the order of those two, causing the constexpr
-    # to be last. See the cpp_function_attributes test for a verification.
-    elif func.type.endswith('constexpr'):
-        func.type = func.type[:-10]
-        func.is_constexpr = True
-    else:
-        func.is_constexpr = False
-    # When 1.8.18 encounters `constexpr static`, it keeps the static there. For
-    # `static constexpr` it doesn't. In both cases the static="yes" is put
-    # there correctly. WHY DOXYGEN, WHY?!
-    if func.type.startswith('static'):
-        func.type = func.type[7:]
+    #
+    # First the prefix keywords - Doxygen has a habit of leaking attributes and
+    # other specifiers into the function's return type, and not necessarily
+    # in any consistent order (including swapping it with the actual type!)
+    func.is_constexpr = False
+    func.is_consteval = False
+    func.is_explicit = False
+    func.is_virtual = False
+    func.is_static = False
+    keywords_behaving_badly = (
+        'constexpr',
+        'consteval',
+        'explicit',
+        'static',
+        'friend',
+        'inline',
+        'virtual'
+    )
+    matched_bad_keyword = True
+    while matched_bad_keyword:
+        matched_bad_keyword = False
+        for kw in keywords_behaving_badly:
+            if func.type == kw: # constructors
+                matched_bad_keyword = True
+                func.type = ''
+            elif func.type.startswith(kw+' '):
+                matched_bad_keyword = True
+                func.type = func.type[len(kw):].strip()
+            elif func.type.endswith(' '+kw):
+                matched_bad_keyword = True
+                func.type = func.type[:len(kw)].strip()
```

Comment:
Hm, this wasn't covered in the original code either, so probably the comment saying

> For some effing reason, when a constexpr function has decltype(auto) return type, Doxygen swaps the order of those two [...]

doesn't hold anymore and this is a bug that's no longer present in 1.8.16+? Assuming you can't reproduce in any of the tested versions (feel free to abuse the CI by force-pushing), I'd drop this branch and also update the `"It's just HORRIBLE."` comment in `cpp_function_attributes` saying that this is no longer an issue on 1.8.16+ (but keeping the `decltype(auto)` function there to catch future regressions).

Reply:
Well the main benefit to structuring it this way is to guard against future doxygen regressions (like, say in 1.9.5 they bring back the old `decltype(auto)` behaviour); assuming C++ never allows these keywords to be used as typenames (which seems pretty certain), scanning for keywords from both sides of the typename will always yield the correct result regardless of what doxygen decides to do.

Reply:
Okay, then an explanatory comment would be the best

```py
# Needed before 1.8.16, currently unused but what if it returns again!
```